File: templates/Uwp/Pages/TreeView.Prism/Param_ProjectName/TemplateSelectors/SampleDataTemplateSelector.cs
using Param_RootNamespace.Core.Models;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;

namespace Param_RootNamespace.TemplateSelectors
{
    public class SampleDataTemplateSelector : DataTemplateSelector
    {
        public DataTemplate CompanyTemplate { get; set; }

        public DataTemplate OrderTemplate { get; set; }

        public DataTemplate OrderDetailTemplate { get; set; }

        protected override DataTemplate SelectTemplateCore(object item)
        {
            return GetTemplate(item) ?? base.SelectTemplateCore(item);
        }

        protected override DataTemplate SelectTemplateCore(object item, DependencyObject container)
        {
            return GetTemplate(item) ?? base.SelectTemplateCore(item, container);
        }

        private DataTemplate GetTemplate(object item)
        {
            switch (item)
            {
                case SampleCompany company:
                    return CompanyTemplate;
                case SampleOrder order:
                    return OrderTemplate;
                case SampleOrderDetail orderDetail:
                    return OrderDetailTemplate;
            }

            return null;
        }
    }
}

File: templates/Uwp/Pages/TreeView._VB/TemplateSelectors/SampleDataTemplateSelector.vb
Imports Param_RootNamespace.Core.Models

Namespace TemplateSelectors
    Public Class SampleDataTemplateSelector
        Inherits DataTemplateSelector

        Public Property CompanyTemplate As DataTemplate

        Public Property OrderTemplate As DataTemplate

        Public Property OrderDetailTemplate As DataTemplate

        Protected Overrides Function SelectTemplateCore(item As Object) As DataTemplate
            If item IsNot Nothing Then
                Return GetTemplate(item)
            End If
            Return MyBase.SelectTemplateCore(item)
        End Function

        Protected Overrides Function SelectTemplateCore(item As Object, container As DependencyObject) As DataTemplate
            If item IsNot Nothing Then
                Return GetTemplate(item)
            End If
            Return MyBase.SelectTemplateCore(item, container)
        End Function

        Private Function GetTemplate(item As Object) As DataTemplate
            Select Case item.GetType()
                Case GetType(SampleCompany)
                    Return CompanyTemplate
                Case GetType(SampleOrder)
                    Return OrderTemplate
                Case GetType(SampleOrderDetail)
                    Return OrderDetailTemplate
            End Select

            Return Nothing
        End Function
    End Class
End Namespace